caas/kubernetes/provider/exec/exec.go

```diff
@@ -0,0 +1,248 @@
+	}
+
+	if ep.PodName, ep.ContainerName, err = getValidatedPodContainer(
+		podGetter, ep.PodName, ep.ContainerName,
+	); err != nil {
+		return errors.Trace(err)
+	}
+	return nil
+}
+
+// Exec runs commands on a pod in the cluster.
+func (c client) Exec(params ExecParams, cancel <-chan struct{}) error {
+	if err := params.validate(c.podGetter); err != nil {
+		return errors.Trace(err)
+	}
+	return errors.Trace(c.exec(params, cancel))
+}
+
+func processEnv(env []string) string {
+	out := ""
+	for _, v := range env {
+		out += fmt.Sprintf("export %s; ", v)
+	}
+	return out
+}
+
+func (c client) exec(opts ExecParams, cancel <-chan struct{}) error {
+	cmd := ""
+	if opts.WorkingDir != "" {
+		cmd += fmt.Sprintf("cd %s; ", opts.WorkingDir)
+	}
+	if len(opts.Env) > 0 {
+		cmd += processEnv(opts.Env)
+	}
+	cmd += fmt.Sprintf("%s; ", strings.Join(opts.Commands, " "))
+	cmdArgs := []string{"sh", "-c", cmd}
+
+	req := c.clientset.CoreV1().RESTClient().Post().
+		Resource("pods").
+		Name(opts.PodName).
+		Namespace(c.namesapce).
+		SubResource("exec").
+		Param("container", opts.ContainerName).
+		VersionedParams(&core.PodExecOptions{
+			Container: opts.ContainerName,
+			Command:   cmdArgs,
+			Stdin:     opts.Stdin != nil,
+			Stdout:    opts.Stdout != nil,
+			Stderr:    opts.Stderr != nil,
+			TTY:       false,
+		}, scheme.ParameterCodec)
+
+	executor, err := c.remoteCmdExecutorGetter("POST", req.URL())
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	errChan := make(chan error, 1)
+	go func() {
+		errChan <- executor.Stream(remotecommand.StreamOptions{
+			Stdin:  opts.Stdin,
+			Stdout: opts.Stdout,
+			Stderr: opts.Stderr,
+			Tty:    false,
+		})
+	}()
+	select {
+	case err := <-errChan:
+		return errors.Trace(err)
+	case <-cancel:
+		return errors.New(fmt.Sprintf("exec cancelled: %v", opts))
+	}
+}
+
+func parsePodName(podName string) (string, error) {
+	err := errors.NotValidf("podName %q", podName)
+	slice := strings.SplitN(podName, "/", 2)
+	if len(slice) == 1 {
+		podName = slice[0]
+	} else if slice[0] != "pod" {
+		return "", err
+	} else {
+		podName = slice[1]
+	}
+	if len(podName) == 0 {
+		return "", err
+	}
+	return podName, nil
+}
+
+func getValidatedPodContainer(
+	podGetter typedcorev1.PodInterface, podName, containerName string,
+) (string, string, error) {
+	var err error
+	if podName, err = parsePodName(podName); err != nil {
+		return "", "", errors.Trace(err)
+	}
+	var pod *core.Pod
+	pod, err = podGetter.Get(podName, metav1.GetOptions{})
+	if k8serrors.IsNotFound(err) {
+		logger.Debugf("no pod named %q found", podName)
+		logger.Debugf("try get pod by UID for %q", podName)
+		pods, err := podGetter.List(metav1.ListOptions{})
+		// TODO(caas): remove getting pod by Id (a bit expensive) once we started to cache podName in cloudContainer doc.
+		if err != nil {
+			return "", "", errors.Trace(err)
+		}
+		for _, p := range pods.Items {
+			if string(p.GetUID()) == podName {
+				podName = p.GetName()
+				pod = &p
```

Comment:
Not really a problem here, but I'm not a fan of taking a ptr of a loop variable.